src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php

```diff
@@ -0,0 +1,358 @@
+                    $value = $serialized;
+                }
+            } elseif (is_string($value)) {
+                // Serialize strings if they could be confused with serialized objects or arrays
+                if ('N;' === $value || (isset($value[2]) && ':' === $value[1])) {
+                    $value = serialize($value);
+                }
+            } elseif (!is_scalar($value)) {
+                throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable %s value', $key, gettype($value)));
+            }
+
+            $dump .= var_export($key, true).' => '.var_export($value, true).",\n";
+        }
+
+        $dump .= "\n);\n";
+        $dump = str_replace("' . \"\\0\" . '", "\0", $dump);
+
+        $tmpFile = uniqid($this->file);
+
+        file_put_contents($tmpFile, $dump);
+        @chmod($tmpFile, 0666);
+        unset($serialized, $unserialized, $value, $dump);
+
+        @rename($tmpFile, $this->file);
+
+        $this->values = (include $this->file) ?: array();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem($key)
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
+
+        if (!is_string($key)) {
+            throw new InvalidArgumentException(sprintf('Cache key must be string, "%s" given', is_object($key) ? get_class($key) : gettype($key)));
+        }
+
+        if (!isset($this->values[$key])) {
+            return $this->fallbackPool->getItem($key);
+        }
+
+        $value = $this->values[$key];
+
+        if ('N;' === $value) {
+            $value = null;
+        } elseif (is_string($value) && isset($value[2]) && ':' === $value[1]) {
+            $value = unserialize($value);
+        }
+
+        $f = $this->createCacheItem;
+
+        return $f($key, $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems(array $keys = array())
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
+
+        foreach ($keys as $key) {
+            if (!is_string($key)) {
+                throw new InvalidArgumentException(sprintf('Cache key must be string, "%s" given', is_object($key) ? get_class($key) : gettype($key)));
+            }
+        }
+
+        return $this->generateItems($keys);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasItem($key)
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
+
+        if (!is_string($key)) {
+            throw new InvalidArgumentException(sprintf('Cache key must be string, "%s" given', is_object($key) ? get_class($key) : gettype($key)));
+        }
+
+        return isset($this->values[$key]) || $this->fallbackPool->hasItem($key);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        $this->values = array();
+
+        @unlink($this->file);
```

Comment:
Let's do by using `file_exists` after unlinking when `unlink` returns `false`. Even if it opens a race condition, warming up the cache should be rare and slower than the time between the `unlink` and `file_exists`, so that in reality the race doesn't happen.